crates/uv-auth/src/store.rs

```diff
@@ -0,0 +1,175 @@
+use lazy_static::lazy_static;
+use std::{collections::HashMap, sync::Mutex};
+
+use netrc::Authenticator;
+use tracing::warn;
+use url::Url;
+
+use crate::NetLoc;
+
+lazy_static! {
+    // Store credentials for NetLoc
+    static ref PASSWORDS: Mutex<HashMap<NetLoc, Option<Credential>>> = Mutex::new(HashMap::new());
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum Credential {
+    Basic(BasicAuthData),
+    UrlEncoded(UrlAuthData),
+}
+
+impl Credential {
+    pub fn username(&self) -> String {
+        match self {
+            Credential::Basic(auth) => auth.username.clone(),
+            Credential::UrlEncoded(auth) => auth.username.clone(),
+        }
+    }
+    pub fn password(&self) -> Option<String> {
+        match self {
+            Credential::Basic(auth) => auth.password.clone(),
+            Credential::UrlEncoded(auth) => auth.password.clone(),
+        }
+    }
+}
+
+impl From<&Authenticator> for Credential {
+    fn from(auth: &Authenticator) -> Self {
+        Credential::Basic(BasicAuthData {
+            username: auth.login.clone(),
+            password: Some(auth.password.clone()),
+        })
+    }
+}
+
+// Used for URL encoded auth in User info
+// <https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.1>
+#[derive(Clone, Debug, PartialEq)]
+pub struct UrlAuthData {
+    pub username: String,
+    pub password: Option<String>,
+}
+
+impl UrlAuthData {
+    pub fn apply_to_url(&self, mut url: Url) -> Url {
+        url.set_username(&self.username)
+            .unwrap_or_else(|()| warn!("Failed to set username"));
+        url.set_password(self.password.as_deref())
+            .unwrap_or_else(|()| warn!("Failed to set password"));
+        url
+    }
+}
+
+// HttpBasicAuth - Used for netrc and keyring auth
+// <https://datatracker.ietf.org/doc/html/rfc7617>
+#[derive(Clone, Debug, PartialEq)]
+pub struct BasicAuthData {
+    pub username: String,
+    pub password: Option<String>,
+}
+
+pub struct AuthenticationStore;
```

Comment:
I'm not a huge fan of the singleton struct here but I don't need to block the PR on it. It's just not something we've really done anywhere else that I can remember. Maybe @BurntSushi would have a better opinion on it.

Reply:
I imagined the passwords would live on it but I think that's infeasible for some reason?

Reply:
My initial plan was for a store instance to be a field of the `AuthMiddleware` struct and the store to hold an instance of the password table - but you do not have access to mutable self in the `Middleware` trait, so that was the limitation I ran into.  Combine that with keeping the copying from trusted URL that wouldn't have access to the same instance - I went with this approach.

Reply:
Can you just use interior mutability to get around this? e.g. `RefCell<T>`?

Reply:
I believe you could just do it with the Mutex within the struct. But we can change this later if desired.